tools/CoqMakefile.in

```diff
@@ -208,8 +209,33 @@ COQEXTRAFLAGS?=
 COQCHKEXTRAFLAGS?=
 COQDOCEXTRAFLAGS?=
 
+# Find the last argument of the form "-native-compiler FLAG"
+COQUSERNATIVEFLAG:=$(strip \
+$(subst -native-compiler-,,\
+$(lastword \
+$(filter -native-compiler-%,\
+$(subst -native-compiler ,-native-compiler-,\
+$(strip $(COQEXTRAFLAGS)))))))
+
+COQFILTEREDEXTRAFLAGS:=$(strip \
+$(filter-out -native-compiler-%,\
+$(subst -native-compiler ,-native-compiler-,\
+$(strip $(COQEXTRAFLAGS)))))
+
+ifneq (,$(COQUSERNATIVEFLAG))
+  COQNATIVEFLAGS="-native-compiler" "$(COQUSERNATIVEFLAG)"
+  COQDONATIVE=$(COQUSERNATIVEFLAG)
+else
+ifeq '$(COQMF_COQ_NATIVE_COMPILER_DEFAULT)' 'yes'
+  COQNATIVEFLAGS="-native-compiler" "ondemand"
+  COQDONATIVE="yes"
+else
+  COQNATIVEFLAGS=
+endif
+endif
+
 # these flags do NOT contain the libraries, to make them easier to overwrite
-COQFLAGS?=-q $(OTHERFLAGS) $(COQEXTRAFLAGS)
+COQFLAGS?=-q $(COQNATIVEFLAGS) $(OTHERFLAGS) $(COQFILTEREDEXTRAFLAGS)
```

Comment:
IIUC rewriter puts `-native-compiler ondemand` in the coqproject which ends up in OTHERFLAGS. Reordering COQNATIVEFLAGS first prevents us from overriding it. Therefore this should be
```suggestion
COQFLAGS?=-q $(OTHERFLAGS) $(COQFILTEREDEXTRAFLAGS) $(COQNATIVEFLAGS)
```
 
or maybe just 
```suggestion
COQFLAGS?=-q $(OTHERFLAGS) $(COQEXTRAFLAGS)
```
, I don't remember why we filter.

Reply:
We filter to prevent passing the user-provided `-native-compiler` option. Even if it's set to `yes`, `coqnative` should be responsible for the native compilation phase, not `coqc`.

Reply:
But doesn't it got right back in through COQNATIVEFLAGS?

Reply:
If I understand correctly, coq-rewriter is not respecting our documentation here. It explicitly states that native compilation options need to be passed through `COQEXTRAFLAGS`, not in the `_CoqProject`. Shouldn't we fix it that way instead of circumventing our hacky users?

Reply:
> doesn't it got right back

You're right. I think we have to make a complete analysis of what is expected depending on the configure flag and the optional `COQEXTRAFLAG`. That's 3 × 4 = 12 possibilities there.

Reply:
> If I understand correctly, coq-rewriter is not respecting our documentation here. It explicitly states that native compilation options need to be passed through `COQEXTRAFLAGS`, not in the `_CoqProject`.

How should a project specify that it's not compatible with `-native-compiler yes` while still enabling the `-native-compiler` and also allowing users to pass in whatever they want in `COQEXTRAFLAGS`?

Reply:
I am not sure I understand the question. Are you asking for an explicit hierarchisation of the various flags we can pass to Coq, or are you advocating for the support of the `-native-compiler` flag in `_CoqProject`?

Reply:
No.  It seems that you're requesting a change to coq-rewriter.  I'm asking for a code snippet and a filename to paste it in to replace the current snippet in _CoqProject, such that it will have the same effect as the current code regardless of how the user sets `COQEXTRAFLAGS`.  (For example, does using `COQEXTRAFLAGS += -native-compiler ondemand` in the Makefile.coq.local always do the right thing?)

Reply:
And, if there is no such code snippet, then, yes, I'd advocate supporting this setting in _CoqProject

Reply:
I added a patch to support natively the `-native-compiler` flag in _CoqProject. As explained, the flag is taken in decreasing priority from the `COQEXTRAFLAGS` variable, the _CoqProject file and then the configure-time option. Even without this change the first patch of this PR is backwards compatible with coq-rewriter, but it's not very robust. Is this fine with you, or do you want another mechanism?

Reply:
That sounds fine.  To be clear, the current set-up will continue to work, but it's recommended that I switch to passing `-native-compiler` without `-arg` in new enough versions of Coq?

Reply:
Yes, it'd be better indeed.